src/group_theory/free_product.lean

```diff
@@ -151,6 +151,36 @@ lemma of_left_inverse [decidable_eq ι] (i : ι) :
 lemma of_injective (i : ι) : function.injective ⇑(of : M i →* _) :=
 by { classical, exact (of_left_inverse i).injective }
 
+lemma lift_range_subset_submonoid {N} [monoid N] (f : Π i, M i →* N) {s : submonoid N}
+  (h : ∀ i, set.range (f i) ⊆ s) : set.range (lift f) ⊆ s :=
+begin
+  rintros _ ⟨x, rfl⟩,
+  induction x using free_product.induction_on with i x x y hx hy,
+  { exact s.one_mem, },
+  { simp only [lift_of, set_like.mem_coe], exact h i (set.mem_range_self x), },
+  { simp only [map_mul, set_like.mem_coe], exact s.mul_mem hx hy, },
+end
+
+lemma range_eq_submonoid_closure {N} [monoid N] (f : Π i, M i →* N) :
```

Comment:
This seems like a strange result to me.

Is it possible that what we should really be proving is:
```lean
  ⨆ i, (f i).mrange = submonoid.closure (⋃ i, set.range (f i))
```

Reply:
I tried to be consistent with https://leanprover-community.github.io/mathlib_docs/group_theory/free_group.html#free_group.lift.range_eq_closure. are you saying that should be changed as well?

Reply:
I see; consistency is certainly good but if it was easy I'd favour achieving this by changing the statements in `free_group`. For example I think the two results you mention should state:

- `(lift f).range ≤ s :=`
- `(lift f).range = subgroup.closure (set.range f) :=`

Do you think you could try making these changes?

If it turns out not to be easy then we can think again.

Reply:
I'll try. These are the statements I had expected myself.